presto-orc/src/main/java/com/facebook/presto/orc/writer/SegmentedSliceBlockBuilder.java

```diff
@@ -0,0 +1,416 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.orc.writer;
+
+import com.facebook.presto.common.block.AbstractVariableWidthBlock;
+import com.facebook.presto.common.block.Block;
+import com.facebook.presto.common.block.BlockBuilder;
+import com.facebook.presto.common.block.BlockBuilderStatus;
+import io.airlift.slice.DynamicSliceOutput;
+import io.airlift.slice.Slice;
+import io.airlift.slice.SliceInput;
+import io.airlift.slice.XxHash64;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.Arrays;
+import java.util.function.BiConsumer;
+
+import static io.airlift.slice.SizeOf.sizeOf;
+import static java.lang.String.format;
+
+/**
+ * Custom Block Builder implementation for use with SliceDictionaryBuilder.
+ * Instead of using one large contiguous Slice for storing the unique Strings
+ * in String dictionary, this class uses Segmented Slices.
+ *
+ * Why implement a block builder ?
+ * SliceDictionaryBuilder takes in a Block and Position to write.
+ *  1. It can create a slice for the position and write it. This does not
+ *      require a block builder. But temporary slice, produces lot of
+ *      short lived garbage.
+ *  2. A block and position can be copied to BlockBuilder using the method
+ *  Block.writeBytesTo . But this requires implementing the BlockBuilder interface.
+ *  Most of the methods are going to be unused and left as Unsupported.
+ *
+ * What's the difference between this class and VariableWidthBlockBuilder?
+ * This class is different from VariableWidthBlockBuilder in the following ways
+ *      1. It does not support nulls. (So null byte array and management is avoided).
+ *      2. Instead of using one contiguous chunk for storing all the entries,
+ *      they are segmented.
+ *
+ * How is it implemented ?
+ * The Strings from 0 to SEGMENT_SIZE-1 are stored in the first segment.
+ * The string from SEGMENT_SIZE to 2 * SEGMENT_SIZE -1 goes to the second.
+ * Each segment has Slice(data is concatenated and stored in one slice)
+ * and offsets to capture the start offset and length. New slices are appended
+ * to the open segment. Once the segment is full the segment is
+ * finalized and appended to the closed segments. A new open segment is
+ * created for further appends.
```

Comment:
Can we elaborate a bit on the benefit of chopping a slice into `Slice[]`? Is it because of performance? If yes, could you help to elaborate why having a continuous slice will be slow?

Reply:
Added the reason to the top. This is mainly for memory, though it regresses CPU little bit.